### PR TITLE
feat: new grid

### PR DESCRIPTION
**Pull request by fangwufeng-v**
### Feature Preview

---

#### PR Checklist

- [ ] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [ ] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+              key={option.id}
+              size="small"
+              color={option.color}
+              label={option.name}
+            />
+          ))}
+        onChange={handleChange}
+        onClose={handleClose}
+      >
+        <ListSubheader className="text-xs mb-4">
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx:113]: How about extact the component to make it a bit cleaner and easier to manager.
```ts
const CustomListSubheader = ({ title, inputProps }) => (
  <ListSubheader className="text-xs mb-4">{title}</ListSubheader>
);

const CustomInputSubheader = ({ value, onInput }) => (
  <ListSubheader className="flex">
    <OutlinedInput size="small" value={value} onInput={onInput} />
  </ListSubheader>
);

<CustomListSubheader title={t('grid.selectOption.createTag')} />
<CustomInputSubheader value={newOptionName} onInput={handleInput} />

```

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        </ListSubheader>
+        <ListSubheader className="flex">
+          <OutlinedInput
+            size="small"
+            value={newOptionName}
+            onInput={handleInput}
+          />
+        </ListSubheader>
+        {newOptionName ? (
+          <MenuItem
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx:124]:
```ts
const NewTagMenuItem = ({ newOptionName, selectedIds, handleNewTagClick, t }) => (
  newOptionName ? (
    <MenuItem
      className="mt-2 text-xs text-neutral-500 font-medium"
      value={selectedIds}
      onClick={handleNewTagClick}
    >
      {t('grid.selectOption.create')}
      <Tag className="ml-2" size="small" label={newOptionName} />
    </MenuItem>
  ) : null
);

const ExistingTagsHeader = ({ options, t }) => (
  options.length ? (
    <ListSubheader className="text-xs mt-4 mb-2">
      {t('grid.selectOption.orSelectOne')}
    </ListSubheader>
  ) : null
);

const ExistingTagMenuItem = ({ option, index, last, handleDetailsClick }) => (
  <MenuItem
    className={index !== last ? 'mb-2' : ''}
    key={option.id}
    value={option.id}
  >
    <div className="flex-1">
      <Tag key={option.id} size="small" color={option.color} label={option.name} />
    </div>
    <IconButton onClick={event => handleDetailsClick(event, option)}>
      <DetailsSvg className="text-base" />
    </IconButton>
  </MenuItem>
);

// Usage
<NewTagMenuItem newOptionName={newOptionName} selectedIds={selectedIds} handleNewTagClick={handleNewTagClick} t={t} />
<ExistingTagsHeader options={options} t={t} />
{options.map((option, index) => (
  <ExistingTagMenuItem option={option} index={index} last={options.length - 1} handleDetailsClick={handleDetailsClick} />
))}

```

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridTable/GridTable.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    horizontal: true,
+    count: fields.length,
+    overscan: 5,
+    getItemKey: i => fields[i].id,
+    getScrollElement: () => horizontalScrollElementRef.current,
+    estimateSize: (i) => fields[i].width ?? defaultWidth,
+  });
+
+  const columnVirtualItems = columnVirtualizer.getVirtualItems();
+  const [before, after] = columnVirtualItems.length > 0
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridTable/GridTable.tsx:62]: How about extract into a single function.
```ts
const calculateBeforeAfter = (columnVirtualItems: VirtualItem[], columnVirtualizer: Virtualizer) => {
  return columnVirtualItems.length > 0
    ? [
        columnVirtualItems[0].start,
        columnVirtualizer.getTotalSize() - columnVirtualItems[columnVirtualItems.length - 1].end,
      ]
    : [0, 0];
};
```

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridTable/GridTable.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+  const horizontalScrollElementRef = useRef<HTMLDivElement>(null);
+
+  const defaultWidth = 221;
+  const defaultHeight = 41;
+
+  const renderRows = useMemo<RenderRow[]>(() => {
+    return [
+      {
+        type: 'fields' as const,
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridTable/GridTable.tsx:19]: How about define a enum here:
```ts
enum RenderType {
  Fields = 'fields',
  Row = 'row'
}

const renderRows = useMemo<RenderRow[]>(() => [
    {
      type: RenderType.Fields,
    },
    ...rows.map(row => ({
      type: RenderType.Row,
      data: row,
    })),
  ], [rows]);
```

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridTable/GridTable.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  const columnVirtualItems = columnVirtualizer.getVirtualItems();
+  const [before, after] = columnVirtualItems.length > 0
+    ? [
+        columnVirtualItems[0].start,
+        columnVirtualizer.getTotalSize() - columnVirtualItems[columnVirtualItems.length - 1].end,
+      ]
+    : [0, 0]
+
+  return (
+    <div
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridTable/GridTable.tsx:70]: Also wrap into a single component:
```ts
const VirtualizedRows: FC<{ /* props */ }> = ({ rowVirtualizer, renderRows, columnVirtualItems, before, after }) => (
  <div style={{ position: 'relative', height: rowVirtualizer.getTotalSize() }}>
    {rowVirtualizer.getVirtualItems().map(({ key, index, start, size }) => {
      const row = renderRows[index];
      const needMeasure = row.type !== RenderType.Row;

      return (
        <div
          ref={needMeasure ? rowVirtualizer.measureElement : undefined}
          key={key}
          className="absolute top-0 left-0 flex min-w-full border-b border-line-divider"
          style={{
            height: needMeasure ? undefined : size,
            transform: `translateY(${start}px)`,
          }}
        >
          <GridRow row={row} columnVirtualItems={columnVirtualItems} before={before} after={after} />
        </div>
      );
    })}
  </div>
);

```

File: frontend/appflowy_tauri/src/appflowy_app/hooks/notification.hooks.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+};
+
+type NotificationMap = typeof NotificationPBMap;
+
+type NotificationEnum = keyof NotificationMap;
+
+type NullableInstanceType<K extends ((abstract new (...args: any) => any) | null)> = K extends (abstract new (...args: any) => any) ? InstanceType<K> : void;
+
+type NotificationHandler<K extends NotificationEnum> = (result: Result<NullableInstanceType<NotificationMap[K]>, FlowyError>) => void;
+
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/hooks/notification.hooks.ts:39]: It should have some documentation here.

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridField/GridField.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        });
+      } catch (e) {
+        // TODO
+        console.error(`change field ${field.id} name from '${field.name}' to ${inputtingName} fail`, e);
+      }
+    }
+  }, [viewId, field, inputtingName]);
+
+  return (
+    <div
```
Comment (appflowy) [inline frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridField/GridField.tsx:72]: It's not easier to understand. Extract into multiple components would be better

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx (hunk 1 above)
Comment (fangwufeng-v) [reply frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx]: Due to the change of the UI draft, there are some logical changes here.

And I think `ListSubheader` and `OutlinedInput` is simple enough so no need to extact components like `CustomListSubheader` or `CustomInputSubheader`.

For cleaner to read, I use clear variable `searchInput` to name the code segment.

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx (hunk 2 above)
Comment (fangwufeng-v) [reply frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridCell/GridSelectCell/GridSelectCell.tsx]: I tried modifying like you wrote, but ran into a problem.

The `Select` of `mui` has a limitation, it **must** use `MenuItem` as a direct child component to implement the selected logic.

So what I can do is keep the `MenuItem` component and wrap the children of `MenuItem` to another functional component called `SelectOptionItem`.

File: frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridField/GridField.tsx (hunk 7 above)
Comment (fangwufeng-v) [reply frontend/appflowy_tauri/src/appflowy_app/components/database/grid/GridField/GridField.tsx:25]: Done. Extract to `GridFieldMenu` and `GridFieldMenuActions`.

File: frontend/appflowy_tauri/src/appflowy_app/hooks/notification.hooks.ts (hunk 6 above)
Comment (fangwufeng-v) [reply frontend/appflowy_tauri/src/appflowy_app/hooks/notification.hooks.ts]: Documentation has been added like that.

by the way, for cleaner, I change a bit code and replace the `NotificationEnum` with `DatabaseNotification`.

**Pull request merged**